File: python/tests/test_shared_fixtures.py
import pytest

from countrycode import countrycode, countryname

from fixture_data import (
    CONVERSIONS_BY_CODE,
    COUNTRYNAME_BY_DESTINATION,
    NAME_NONMATCHES_BY_CODE,
    NAME_VARIATIONS_BY_EXPECTED,
)


NAME_VARIATIONS = [
    (variation, expected)
    for expected, variations in NAME_VARIATIONS_BY_EXPECTED.items()
    for variation in variations
]
CONVERSIONS = [
    (source, origin, destination, expected)
    for origin, destinations in CONVERSIONS_BY_CODE.items()
    for destination, mapping in destinations.items()
    for source, expected in mapping.items()
]
NAME_NONMATCHES = [
    (source, origin, destination)
    for origin, destinations in NAME_NONMATCHES_BY_CODE.items()
    for destination, sources in destinations.items()
    for source in sources
]
COUNTRYNAME_CASES = [
    (source, destination, expected)
    for destination, mapping in COUNTRYNAME_BY_DESTINATION.items()
    for source, expected in mapping.items()
]


@pytest.mark.parametrize(
    ("variation", "expected"),
    NAME_VARIATIONS,
)
def test_known_name_variations(variation, expected):
    actual = countrycode(
        variation,
        "country.name",
        "country.name",
        warn=False,
    )
    assert actual == expected


@pytest.mark.parametrize(
    ("source", "origin", "destination", "expected"),
    CONVERSIONS,
)
def test_shared_conversion_cases(source, origin, destination, expected):
    actual = countrycode(source, origin, destination, warn=False)
    assert actual == expected


@pytest.mark.parametrize(
    ("variation", "origin", "destination"),
    NAME_NONMATCHES,
)
def test_shared_name_nonmatches(variation, origin, destination):
    actual = countrycode(variation, origin, destination, warn=False)
    assert actual is None


@pytest.mark.parametrize(
    ("source", "destination", "expected"),
    COUNTRYNAME_CASES,
)
def test_shared_countryname_cases(source, destination, expected):
    assert countryname(source, destination, warn=False) == expected

File: r/tests/testthat/test-shared-fixtures.R
test_that("shared name variations match", {
  expected_names <- read_shared_fixture("name-variations.yaml")

  for (expected in names(expected_names)) {
    for (source in expected_names[[expected]]) {
      actual <- countrycode(
        source,
        origin = "country.name",
        destination = "country.name",
        warn = FALSE
      )
      expect_equal(actual, expected)
    }
  }
})

test_that("shared conversion cases match", {
  origins <- read_shared_fixture("conversions.yaml")

  for (origin in names(origins)) {
    for (destination in names(origins[[origin]])) {
      cases <- origins[[origin]][[destination]]
      for (source in names(cases)) {
        source_value <- source
        if (origin %in% names(codelist) && is.numeric(codelist[[origin]])) {
          source_value <- as.integer(source)
        }
        actual <- countrycode(
          source_value,
          origin = origin,
          destination = destination,
          warn = FALSE
        )
        expected <- cases[[source]]
        if (is.null(expected)) {
          expect_true(is.na(actual))
        } else {
          expect_equal(actual, expected)
        }
      }
    }
  }
})

test_that("shared name nonmatches remain missing", {
  origins <- read_shared_fixture("name-nonmatches.yaml")

  for (origin in names(origins)) {
    for (destination in names(origins[[origin]])) {
      for (source in origins[[origin]][[destination]]) {
        actual <- countrycode(
          source,
          origin = origin,
          destination = destination,
          warn = FALSE
        )
        expect_true(is.na(actual))
      }
    }
  }
})

test_that("shared countryname cases match", {
  destinations <- read_shared_fixture("countryname.yaml")

  for (destination in names(destinations)) {
    cases <- destinations[[destination]]
    for (source in names(cases)) {
      actual <- countryname(source, destination = destination, warn = FALSE)
      expected <- cases[[source]]
      if (is.null(expected)) {
        expect_true(is.na(actual))
      } else {
        expect_equal(actual, expected)
      }
    }
  }
})
